akida: reset index if a different hardware variant is found

rchevalier-brainchip opened the pull request:
Currently when one AKD1000 and one AKD1500 are plugged in the following is seen:

  /dev/akida0
  /dev/akd1500_1

Changes aims at presenting:

  /dev/akida0
  /dev/akd1500_0

hcodina commented on akida-pcie-core.c:
```diff
@@ -1090,7 +1093,8 @@ static int akida_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 	akida->miscdev.name = devm_kasprintf(&pdev->dev, GFP_KERNEL,
 					     "%s%d", ops.miscdev_name,
-					     akida->devno);
+					     ops.index);
```
Don't do that for two reasons:
- The number (index) will not be released when the device is removed
- The incrementation is not atomic

If you really want to have the number based on the board familly, use two different akida_devno (one per board family) and use the correct one according to the board.
Someting like this:
```
diff --git a/akida-pcie-core.c b/akida-pcie-core.c
index de74ef3..a1bd00a 100644
--- a/akida-pcie-core.c
+++ b/akida-pcie-core.c
@@ -25,7 +25,8 @@
 #include "dw-edma-core.h"
 #include "akida-edma.h"
 
-static DEFINE_IDA(akida_devno);
+static DEFINE_IDA(akida_1000_devno);
+static DEFINE_IDA(akida_1500_devno);
 
 /* The DMA RAM area contains eDMA linked-list (LL) and data (DT).
  * This area is used by the eDMA controler and is located inside the device.
@@ -85,6 +86,7 @@ struct akida_dma_chan {
 
 struct akida_dev {
 	struct pci_dev *pdev;
+	struct ida *ida;
 	int devno;
 	struct miscdevice miscdev;
 	struct dw_edma_chip edma_chip;
@@ -836,23 +838,25 @@ static int akida_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 	struct akida_ops ops;
 	int ret, nr_irqs;
 
+	akida = devm_kzalloc(&pdev->dev, sizeof(*akida), GFP_KERNEL);
+	if (!akida)
+		return -ENOMEM;
+
+	akida->pdev = pdev;
+
 	switch (board_id) {
 	case AKIDA_1000:
 		ops = akida_1000_ops;
+		akida->ida = &akida_1000_devno;
 		break;
 	case AKIDA_1500:
 		ops = akida_1500_ops;
+		akida->ida = &akida_1500_devno;
 		break;
 	default:
 		return -EOPNOTSUPP;
 	}
 
-	akida = devm_kzalloc(&pdev->dev, sizeof(*akida), GFP_KERNEL);
-	if (!akida)
-		return -ENOMEM;
-
-	akida->pdev = pdev;
-
 	/* Disable ASPM L0s and L1 states as they cause device stop working */
 	pci_disable_link_state(pdev, PCIE_LINK_STATE_L0S | PCIE_LINK_STATE_L1);
 
@@ -1072,13 +1076,13 @@ static int akida_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 	init_waitqueue_head(&akida->wq_txchan);
 
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
-	ret = ida_simple_get(&akida_devno, 0, 0, GFP_KERNEL);
+	ret = ida_simple_get(akida->ida, 0, 0, GFP_KERNEL);
 	if (ret < 0) {
 		pci_err(pdev, "ida simple get failed (%d)\n", ret);
 		goto fail_akida_dma_exit;
 	}
 #else
-	ret = ida_alloc(&akida_devno, GFP_KERNEL);
+	ret = ida_alloc(akida->ida, GFP_KERNEL);
 	if (ret < 0) {
 		pci_err(pdev, "ida alloc failed (%d)\n", ret);
 		goto fail_akida_dma_exit;
@@ -1108,9 +1112,9 @@ static int akida_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 
 fail_ida_alloc:
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
-	ida_simple_remove(&akida_devno, akida->devno);
+	ida_simple_remove(akida->ida, akida->devno);
 #else
-	ida_free(&akida_devno, akida->devno);
+	ida_free(akida->ida, akida->devno);
 #endif
 fail_akida_dma_exit:
 	akida_dma_exit(akida);
@@ -1128,9 +1132,9 @@ static void akida_remove(struct pci_dev *pdev)
 
 	misc_deregister(&akida->miscdev);
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4, 19, 0)
-	ida_simple_remove(&akida_devno, akida->devno);
+	ida_simple_remove(akida->ida, akida->devno);
 #else
-	ida_free(&akida_devno, akida->devno);
+	ida_free(akida->ida, akida->devno);
 #endif
 	if (akida->txchan[0].chan && akida->rxchan[0].chan)
 		akida_dma_exit(akida);
```
Note that I didn't test this code.

Best regards,
Hervé

rchevalier-brainchip replied:
Sure will modify it and test it tomorrow :) thx !

rchevalier-brainchip replied:
Thanks for the change and details @hcodina ! Confirming that it works as expected (w/ 1 or 2 AKD1000 and 1 AKD1000 + AKD1500)

rchevalier-brainchip committed:
akida: make device id depending on board familly

Currently when one AKD1000 and one AKD1500 are plugged
in the following is seen:

  /dev/akida0
  /dev/akd1500_1

Changes aims at presenting:

  /dev/akida0
  /dev/akd1500_0

Signed-off-by: Richard Chevalier <rchevalier@brainchip.com>

rchevalier-brainchip commented:
If no objection we will merge this change today. Thx!
